x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/utils/get_current_rule_statuses.ts

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { chunk } from 'lodash';
+import { initPromisePool } from '../../../../../utils/promises_pool';
+import { GetCurrentStatusBulkResult, IRuleExecutionLogClient } from '../../../rule_execution_log';
+
+const RULES_PER_CHUNK = 1000;
+
+/**
+ * Get the most recent execution status for each of the given rule IDs.
+ * This method splits work into chunks so not to owerwhelm Elasticsearch
+ * when fetching statuses for a big number of rules.
+ *
+ * @param ruleIds Rule IDs to fetch statuses for
+ * @param execLogClient RuleExecutionLogClient
+ * @param spaceId Current Space ID
+ * @returns A dict with rule IDs as keys and rule statuses as values
+ *
+ * @throws AggregateError if any of the rule status requests fail
+ */
+export async function getCurrentRuleStatuses(
+  ruleIds: string[],
+  execLogClient: IRuleExecutionLogClient,
+  spaceId: string
+): Promise<GetCurrentStatusBulkResult> {
+  const { results, errors } = await initPromisePool({
+    concurrency: 1,
+    items: chunk(ruleIds, RULES_PER_CHUNK),
+    executor: (ruleIdsChunk) =>
+      execLogClient.getCurrentStatusBulk({
+        ruleIds: ruleIdsChunk,
+        spaceId,
+      }),
+  });
```

Comment:
Could we encapsulate this logic in the implementation of `IRuleExecutionLogClient` itself?

Reply:
I thought about that and am not sure we should. It doesn't feel like the exec log client's responsibility and could unnecessarily overcomplicate it. If we look at other clients (SO client, rules client, ES client, etc.), they don't have built-in request chunking as it's needed only for specific use cases. And at this point, it's hard to say how often we would need chunking.

If we start using this method a lot across our codebase, then yes, we can encapsulate that logic. Anyway, it seems pretty trivial to me, so let's postpone the decision for now.

Reply:
I don't have a strong opinion on that, although for me it looks like an implementation details which I'd just hide to reduce complexity of using `IRuleExecutionLogClient`. Let's postpone the decision 👍